### PR TITLE
chore(systemaddon): Use a single method to dispatch TopSites icon stats.

### LINKED ISSUE #3851: Extract TopSites telemetry call into a single method
https://github.com/mozilla/activity-stream/pull/3652#discussion_r151012061

### PR DESCRIPTION
**Pull request by piatra**
Closes #3851

### CODE REVIEW

File: system-addon/content-src/components/TopSites/TopSites.jsx [hunk 1]
```diff
@@ -41,7 +41,10 @@ function countTopSitesIconsTypes(topSites) {
 }
 
 class TopSites extends React.PureComponent {
-  componentDidUpdate() {
+  /**
+   * Dispatch session statistics about the quality of TopSites icons.
+   */
+  _storeTopSitesIconStats() {
```
Comment (Mardak) [inline system-addon/content-src/components/TopSites/TopSites.jsx]: "Store" has an existing usage/meaning, so let's use either "save" or "dispatch"

File: system-addon/content-src/components/TopSites/TopSites.jsx [hunk 2]
```diff
@@ -41,7 +41,10 @@ function countTopSitesIconsTypes(topSites) {
...
     const realTopSites = this.props.TopSites.rows.slice(0, this.props.TopSitesCount);
```
Comment (Mardak) [inline system-addon/content-src/components/TopSites/TopSites.jsx]: What does "real" mean here? Are the ones sliced away not real? If it's supposed to be the same functionality as that of `render`, then let's refactor that.

File: system-addon/test/unit/content-src/components/TopSites.test.jsx [hunk 3]
```diff
@@ -76,33 +76,31 @@ describe("<TopSites>", () => {
...
+      assert.calledOnce(storeStatsSpy);
     });
-    it("should dispatch session perf data on update", () => {
+    it("should call _storeTopSitesIconStats on componentDidUpdate", () => {
       wrapper.instance().componentDidUpdate(DEFAULT_PROPS);
 
+      assert.calledOnce(storeStatsSpy);
+    });
+    it("should dispatch SAVE_SESSION_PERF_DATA", () => {
+      wrapper.instance()._storeTopSitesIconStats(DEFAULT_PROPS);
```
Comment (Mardak) [inline system-addon/test/unit/content-src/components/TopSites.test.jsx]: This method doens't take any parameters, so passing in an argument is confusing

File: system-addon/content-src/components/TopSites/TopSites.jsx (hunk 2 above)
Comment (piatra) [reply system-addon/content-src/components/TopSites/TopSites.jsx]: The way I would do this is set `this.realTopSites` in the `render` method. Alternatively there is `componentWillUpdate` but docs say:
> This method is not called for the initial render.

Comment (Mardak) [reply system-addon/content-src/components/TopSites/TopSites.jsx]: Saving the property is almost like setting state, which shouldn't be done from `render`. And I believe the react pattern is to not set state based on props and instead just provide a helper to compute the value from props.

Comment (Mardak) [reply system-addon/content-src/components/TopSites/TopSites.jsx]: Also, still doesn't answer what "real" means here.

Comment (piatra) [reply system-addon/content-src/components/TopSites/TopSites.jsx]: That was the name used in the TopSites render method. We have two sets of top sites: the ones you regularly see and the ones in the top sites edit modal (that’s where the distinction came from I think). @r1cky knows more about this.

**Comment by as-pine-proxy** [comment]
[Treeherder: aadfc4db1c7252184e69d4d6d5d3e88ca7e4882b](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=aadfc4db1c7252184e69d4d6d5d3e88ca7e4882b)

**Comment by as-pine-proxy** [comment]
[Treeherder: 1b14fb4f6cc34daf8dbf66a7a8acf3779ae4074f](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=1b14fb4f6cc34daf8dbf66a7a8acf3779ae4074f)

**Pull request merged**